ckan/lib/helpers.py

```diff
@@ -200,43 +202,156 @@ def get_site_protocol_and_host():
+    '''Return the URL for an endpoint given some parameters.
 
-    Usage::
+    This is a wrapper for :py:func:`flask.url_for`
+    and :py:func:`routes.url_for` that adds some extra features that CKAN
+    needs.
 
-        import ckan.plugins.toolkit as toolkit
+    To build a URL for a Flask view, pass the name of the blueprint and the
+    view function separated by a period ``.``, plus any URL parameters::
+
+        url_for('api.action', ver=3, logic_function='status_show')
+        # Returns /api/3/action/status_show
+
+    For a fully qualified URL pass the ``_external=True`` parameter. This
+    takes the ``ckan.site_url`` and ``ckan.root_path`` settings into account::
+
+        url_for('api.action', ver=3, logic_function='status_show',
+                _external=True)
+        # Returns http://example.com/api/3/action/status_show
+
+    URLs built by Pylons use the Routes syntax::
 
-        url = toolkit.url_for(controller='package', action='read',
-                              id='my_dataset')
-        => returns '/dataset/my_dataset'
+        url_for(controller='package', action='read', id='my_dataset')
+        # Returns '/dataset/my_dataset'
 
     Or, using a named route::
 
-        toolkit.url_for('dataset_read', id='changed')
+        url_for('dataset_read', id='changed')
+        # Returns '/dataset/changed'
 
-    This is a wrapper for :py:func:`routes.url_for` that adds some extra
-    features that CKAN needs.
+    Use ``qualified=True`` for a fully qualified URL when targeting a Pylons
+    endpoint.
 
+    For backwards compatibility, an effort is made to support the Pylons syntax
+    when building a Flask URL, but this support might be dropped in the future,
+    so calls should be updated.
     '''
+    # Get the actual string code for the locale
     locale = kw.pop('locale', None)
     if locale and isinstance(locale, i18n.Locale):
         locale = i18n.get_identifier_from_locale_class(locale)
+
     # remove __ckan_no_root and add after to not pollute url
     no_root = kw.pop('__ckan_no_root', False)
-    # routes will get the wrong url for APIs if the ver is not provided
-    if kw.get('controller') == 'api':
+
+    # All API URLs generated should provide the version number
+    if kw.get('controller') == 'api' or args and args[0].startswith('api.'):
         ver = kw.get('ver')
         if not ver:
-            raise Exception('api calls must specify the version! e.g. ver=3')
-        # fix ver to include the slash
-        kw['ver'] = '/%s' % ver
-    if kw.get('qualified', False):
-        kw['protocol'], kw['host'] = get_site_protocol_and_host()
-    my_url = _routes_default_url_for(*args, **kw)
+            raise Exception('API URLs must specify the version (eg ver=3)')
+
+    try:
+        # First try to build the URL with the Flask router, making a copy of
+        # the params in case they are modified
+        flask_args = tuple(args)
+        flask_kw = kw.copy()
+
```

Comment:
@smotornyuk `args` and `kw` are the arguments and key words that `h.url_for()` gets. When passing them to `_url_for_flask()` they might get changed, eg trying to adapt them to what Flask actually understands from the Pylons-based syntax (eg [here](https://github.com/ckan/ckan/blob/0581d42a535e7c3bf0a0e9768b58d3be3db91ef6/ckan/lib/helpers.py#L298) and [here](https://github.com/ckan/ckan/blob/0581d42a535e7c3bf0a0e9768b58d3be3db91ef6/ckan/lib/helpers.py#L307), that's why I pass copies of them to `_url_for_flask`:

```python
     my_url = _url_for_flask(*flask_args, **flask_kw)
```

If for some reason Flask can't build the URL, `_url_for_routes` is called with the original args and kw. Does this make sense?